docs/readmes/proposals/p006_subscriber_state_view.md

```diff
@@ -0,0 +1,156 @@
+* For CWF only, we also have an additional entry in DirectoryD that maps from IMSI→UE IPv4 added by PipelineD.
+
+### MobilityD (IMSI+APN → IPv4)
+
+* MobilityD also reports internal state to the orc8r from which we parse out IMSI+APN ↔ IPv4 mapping
+
+
+
+## Summary of currently available subscriber/policy config
+
+For non-Federated deployments, all subscriber and subscriber->policy configurations are made on the Orc8r. So the Orc8r at least has a sense of the desired state.
+For federated deployments where the subscriber configurations are federated to HSS or the policy configurations are federated to PCRF, the Orc8r does not have any visibility into the configurations received.
+
+
+## General Goal
+
+The goal here is to export consolidated session state owned by SessionD to the Orc8r so that we can 
+
+1. View all active session states, uniquely identified by IMSI+APN 
+2. View runtime state of active policies + QoS
+
+
+
+## Proposed State from SessionD → GW State service
+
+Here is the state Protobuf type:
+
+```
+message State {
+    // Type determines how the value is deserialized and validated on the cloud service side
+    string type = 1; // 'subscriber_state'
+    string deviceID = 2; // IMSI'
+    // Value contains the operational state json-serialized.
+    bytes value = 3; // JSON serialized Subscriber State type below
+    uint64 version = 4; // ?
+}
+```
+
+Here is the blob value:
+
+```
+  subscriber_state:
+    type: object
+    description: An object that maps IMSI->APN->SubscriberState
+    properties:
+      imsi:
+        type: string
+      sessions:
+        type: array
+        items:
+          $ref: '#/definitions/policy_rule'
+  session_state:
+    type: object
+    description: Used to describe a session's state
+    properties:
+      apn:
+        type: string
+      msisdn:
+        type: string
+      session_id:
+        type: string
+      ipv4:
+        type: string
+      active_policy_rules:
+        type: array
+        items:
+          $ref:# A JSON-serialized version of protobuf PolicyRule
+```
+
+### Notes on the subscriber_state fields
+
+* IMSI+APN usually uniquely identifies each active session
+    * In SessionD there could be situations where we multiple sessions for the same IMSI+APN at the same time, but there is only one ACTIVE session at a time. Additionally, we do not keep around non-ACTIVE sessions for longer than 5 seconds typically. For simplicity, we can only export ACTIVE sessions so that we don’t have to expose session FSM state to the Orc8r.
+* MSISDN value is propagated from MME to SessionD on session creation. The value is fetched from HSS/SubscriberDB. 
+* This is the first round of fields suggested, but the following fields could be useful for debugging as well
+    * Attach time & session duration in seconds
```

Comment:
Can we add this as well ? Sounds pretty useful

Reply:
Ok yeah will add :D